### DISCUSSION TITLE
Templated Variable not working properly on custom operator

### DISCUSSION THREAD

**Discussion by inception-cto** [apache/airflow#19655] [Q&A] [answered]
**Airflow**: 2.2.2 and 2.2.1

I am having hard time getting the template fields and macros working on my custom operator. I am using Airflow 2.2.1 and have the following code:

```python
S3_KEY="""custom/{{ dag_run.logical_date.strftime('%Y') }}/{{ dag_run.logical_date.strftime('%m') }}/{{ dag_run.logical_date.strftime('%d') }}/custom_test_{{ ts_nodash }}"""

@dag('custom_import_s3',
     default_args=default_args,
     schedule_interval=None,
     start_date=datetime(2021, 1, 1),
     catchup=False,
     tags=['airflow'])
def taskflow_api_etl():
    @task()
    def extract_accounts(**context):
        my_extract = CustomToS3Operator(
                task_id='extract_to_s3',
                s3_bucket=S3_BUCKET,
                s3_key=S3_KEY,
                s3_conn_id=S3_CONNECTION,)
        return my_extract.execute(context)
```
However, when I run the DAG, the S3 key is not templated and the path in S3 is `custom/{{ dag_run.logical_date.strftime('%Y') }}/{{ dag_run.logical_date.strftime('%m') }}/{{ dag_run.logical_date.strftime('%d') }}/custom_test_{{ ts_nodash }}.` 

I have included the template_fields in my custom CustomToS3Operator and I am not sure what I am missing. I tried {{ ts }} and {{ ds }} without much success.

```python
class CustomToS3Operator(BaseOperator):

    template_fields = ('s3_key')

    @apply_defaults
    def __init__(self,
                 s3_conn_id: str,
                 s3_bucket: str,
                 s3_key: str,
                 *args: Any,
                 **kwargs: Any):

        super().__init__(*args, **kwargs)

        self.s3_conn_id = s3_conn_id
        self.s3_bucket = s3_bucket
        self.s3_key = s3_key
```

Any help is appreciated!

**Comment by josh-fell** [comment, marked as answer]
A couple things:
- The `template_fields` attribute for `CustomToS3Operator` needs to be an iterable type. You need to add a comma after "s3_key" for the value to be a tuple.
- You don't need to (and really shouldn't) wrap an operator in a TaskFlow function; just call the operator as you would in the "classic" Airflow use. If you wrap the operator inside a TaskFlow function, you will need to pass `s3_key` as an arg into the TaskFlow function and then pass it to the operator for the templatization to work as expected. The template fields will not be rendered when `CustomToS3Operator` is executed but when `extract_accounts()` is. But again, best practice would be to call the operator directly.

```python
class CustomToS3Operator(BaseOperator):

    template_fields = ('s3_key',)

    def __init__(self,
                 s3_conn_id: str,
                 s3_bucket: str,
                 s3_key: str,
                 *args: Any,
                 **kwargs: Any):

        super().__init__(*args, **kwargs)

        self.s3_conn_id = s3_conn_id
        self.s3_bucket = s3_bucket
        self.s3_key = s3_key

   ...

S3_KEY="""custom/{{ dag_run.logical_date.strftime('%Y') }}/{{ dag_run.logical_date.strftime('%m') }}/{{ dag_run.logical_date.strftime('%d') }}/custom_test_{{ ts_nodash }}"""

@dag('custom_import_s3',
     default_args=default_args,
     schedule_interval=None,
     start_date=datetime(2021, 1, 1),
     catchup=False,
     tags=['airflow'])
def taskflow_api_etl():
    my_extract = CustomToS3Operator(
            task_id='extract_to_s3',
            s3_bucket=S3_BUCKET,
            s3_key=S3_KEY,
            s3_conn_id=S3_CONNECTION,
    )

...

```

**Reply by inception-cto** [reply]
This is super helpful. Thanks! It worked fine.

I used the nested operator to have a better DAG description but with your answer, I changed the DAG linkage to the following:

```python
@dag('custom_import_s3',
     default_args=default_args,
     schedule_interval=None,
     start_date=datetime(2021, 1, 1),
     catchup=False,
     tags=['airflow'])
def taskflow_api_etl():

    my_extract = CustomToS3Operator(
            task_id='extract_to_s3',
            s3_bucket=S3_BUCKET,
            s3_key=S3_KEY,
            s3_conn_id=S3_CONNECTION,
    )

    @task()
    def load(extraction_result):
        """
        #### Load task
        A simple Load task which takes in the result of the Transform task and
        instead of saving it to end user review, just prints it out.
        """

        print(f"Total order value is ok! {extraction_result}")

    my_extract.set_upstream(load(my_extract))

```

Is the final line accurate `my_extract.set_upstream(load(my_extract))`?

**Reply by josh-fell** [reply]
Almost! For using the output from an operator you can use the `.output` property that's available for all operators (see reference [here](https://airflow.apache.org/docs/apache-airflow/stable/tutorial_taskflow_api.html#consuming-xcoms-with-decorated-tasks-from-regular-tasks)).

The last line should be simply:
```python
load(my_extract.output)
```
👆
The beauty of TaskFlow and AIP-31.
